File: jupyter/experiments/clb.py
import os
import fcntl
import mmap

import numpy as np

class clb (object):
    channels_adc = [0, 1]
    channels_dac = [0, 1]

    DWA = 16
    DWG = 14
    DWAr = 2**DWA - 1
    DWGr = 2**DWG - 1
    DWA1 = 2**(DWA-2)
    DWG1 = 2**(DWG-2)
    
    MAGIC = 0xAABBCCDD
    eeprom_device = "/sys/bus/i2c/devices/0-0050/eeprom"
    eeprom_offset = 0x0008

    # FPGA regset structure
    regset_channel_dtype = np.dtype([
        ('ctl_mul', 'int32'),  # multiplication
        ('ctl_sum', 'int32')   # summation
    ])
    regset_dtype = np.dtype([
        ('adc', regset_channel_dtype, 2),  # oscilloscope
        ('dac', regset_channel_dtype, 2),  # generator
    ])

    # floating point structure
    clb_channel_dtype = np.dtype([
        ('gain'  , 'float32'),  # multiplication
        ('offset', 'float32')   # summation
    ])
    clb_range_dtype = np.dtype([
        ('lo', clb_channel_dtype),  #  1.0V range
        ('hi', clb_channel_dtype)   # 20.0V range
    ])
    clb_dtype = np.dtype([
        ('adc', clb_range_dtype  , 2),  # oscilloscope
        ('dac', clb_channel_dtype, 2),  # generator
    ])

    # EEPROM structure
    eeprom_dtype = np.dtype([
        ('adc_hi_gain'  , 'uint32', 2),
        ('adc_lo_gain'  , 'uint32', 2),
        ('adc_lo_offset',  'int32', 2),
        ('dac_gain'     , 'uint32', 2),
        ('dac_offset'   ,  'int32', 2),
        ('magic'        , 'uint32'),
        ('adc_hi_offset',  'int32', 2)
    ])

    def __init__ (self, uio:str = '/dev/uio/clb'):

        # open device file
        try:
            self.uio_dev = os.open(uio, os.O_RDWR | os.O_SYNC)
        except OSError as e:
            raise IOError(e.errno, "Opening {}: {}".format(uio, e.strerror))

        # exclusive lock
        try:
            fcntl.flock(self.uio_dev, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except IOError as e:
            raise IOError(e.errno, "Locking {}: {}".format(uio, e.strerror))

        # map regset
        try:
            self.uio_mem = mmap.mmap(
                fileno=self.uio_dev, length=mmap.PAGESIZE, offset=0x0,
                flags=mmap.MAP_SHARED, prot=(mmap.PROT_READ | mmap.PROT_WRITE))
        except OSError as e:
            raise IOError(e.errno, "Mapping {}: {}".format(uio, e.strerror))

        regset_array = np.recarray(1, self.regset_dtype, buf=self.uio_mem)
        self.regset = regset_array[0]

        tmp_array = np.recarray(1, self.clb_dtype)
        self.tmp = tmp_array[0]

    def __del__ (self):
        self.uio_mem.close()
        try:
            os.close(self.uio_dev)
        except OSError as e:
            raise IOError(e.errno, "Closing {}: {}".format(uio, e.strerror))

    def get_adc_gain (self, ch: int) -> float:
        return (self.regset.adc[ch].ctl_mul / self.DWA1)

    def set_adc_gain (self, ch: int, gain: float):
        self.regset.adc[ch].ctl_mul = int(gain * self.DWA1)

    def get_adc_offset (self, ch: int) -> float:
        return (self.regset.adc[ch].ctl_sum / self.DWAr)

    def set_adc_offset (self, ch: int, offset: float):
        self.regset.adc[ch].ctl_mul = int(offset * self.DWAr)

    def get_dac_gain (self, ch: int) -> float:
        return (self.regset.dac[ch].ctl_mul / self.DWG1)

    def set_dac_gain (self, ch: int, gain: float):
        self.regset.dac[ch].ctl_mul = int(gain * self.DWG1)

    def get_dac_offset (self, ch: int) -> float:
        return (self.regset.dac[ch].ctl_sum / self.DWGr)

    def set_dac_offset (self, ch: int, offset: float):
        self.regset.dac[ch].ctl_mul = int(offset * self.DWGr)

    def FullScaleToVoltage(self, cnt: int) -> float:
        if cnt == 0:
            return (1.0)
        else:
            return (cnt * 100.0 / (1<<32))

    def FullScaleFromVoltage(self, voltage: float) -> int:
        return (int(voltage / 100.0 * (1<<32)));

    def eeprom_read (self):
        # open EEPROM device
        try:
            eeprom_file = open(self.eeprom_device, 'rb')
        except OSError as e:
            raise IOError(e.errno, "Opening {}: {}".format(uio, e.strerror))

        # seek to calibration data
        try:
            eeprom_file.seek(self.eeprom_offset)
        except IOError as e:
            raise IOError(e.errno, "Seek {}: {}".format(uio, e.strerror))

        try:
            buffer = eeprom_file.read (self.eeprom_dtype.itemsize)
        except IOError as e:
            raise IOError(e.errno, "Read {}: {}".format(uio, e.strerror))

        try:
            eeprom_file.close()
        except IOError as e:
            raise IOError(e.errno, "Close {}: {}".format(uio, e.strerror))

        # map buffer onto structure
        eeprom_array = np.recarray(1, self.eeprom_dtype, buf=buffer)
        eeprom_struct = eeprom_array[0]
    
        # missing magic number means a deprecated EEPROM structure was still not updated
        if (eeprom_struct.magic != self.MAGIC):
            for ch in self.channels_adc:
                eeprom_struct.adc_hi_off[ch] = eeprom_struct.adc_lo_off[ch];

        # convert EEPROM values into local float values
        for ch in self.channels_adc:
            self.tmp.adc[ch].lo.gain   = self.FullScaleToVoltage (eeprom_struct.adc_lo_gain[ch]) / 20.0
            self.tmp.adc[ch].hi.gain   = self.FullScaleToVoltage (eeprom_struct.adc_hi_gain[ch])
            self.tmp.adc[ch].lo.offset = eeprom_struct.adc_lo_offset[ch] / (2**13-1)
            self.tmp.adc[ch].hi.offset = eeprom_struct.adc_hi_offset[ch] / (2**13-1) * 20.0
        for ch in self.channels_dac:
            self.tmp.dac[ch].gain   = self.FullScaleToVoltage (eeprom_struct.dac_gain  [ch])
            self.tmp.dac[ch].offset = eeprom_struct.dac_offset[ch] / (2**13-1)

        self.eeprom_struct = eeprom_struct

    def show_float (self):
        for ch in self.channels_adc:
            print('adc[{}].lo.gain   = {}'.format(ch, self.tmp.adc[ch].lo.gain))
            print('adc[{}].hi.gain   = {}'.format(ch, self.tmp.adc[ch].hi.gain))
            print('adc[{}].lo.offset = {}'.format(ch, self.tmp.adc[ch].lo.offset))
            print('adc[{}].hi.offset = {}'.format(ch, self.tmp.adc[ch].hi.offset))
        for ch in self.channels_dac:
            print('dac[{}].gain   = {}'.format(ch, self.tmp.dac[ch].gain))
            print('dac[{}].offset = {}'.format(ch, self.tmp.dac[ch].offset))

File: api1/src/clb.c
#include <math.h>

#include "redpitaya/fixp.h"
#include "redpitaya/clb.h"

int rp_clb_init (rp_clb_t *handle) {
    const char path [] = "/dev/uio/clb";
    
    int status = rp_uio_init (&handle->uio, path);
    if (status) {
        return (-1);
    }
    handle->regset = (rp_clb_regset_t *) handle->uio.map[0].mem;

    // TODO: generalize this, the init function should probably accept types
    // TODO: ADC and DAC handles should be allocated first
    handle->num_dac = 2;
    handle->num_adc = 2;
    int bits_dac = 14;
    int bits_adc = 16;

    // DAC configuration
    for (int unsigned i=0; i<handle->num_dac; i++) {
        const fixp_t mul_t = {.s = 1, .m = 1, .f = bits_dac-2};
        const fixp_t sum_t = {.s = 1, .m = 0, .f = bits_dac-1};
        rp_lin_init(&handle->dac[i], &handle->regset->dac[i], mul_t, sum_t);
    }
    // ADC configuration
    for (int unsigned i=0; i<handle->num_adc; i++) {
        const fixp_t mul_t = {.s = 1, .m = 1, .f = bits_adc-2};
        const fixp_t sum_t = {.s = 1, .m = 0, .f = bits_adc-1};
        rp_lin_init(&handle->adc[i], &handle->regset->adc[i], mul_t, sum_t);
    }

    return(0);
}

int rp_clb_release (rp_clb_t *handle) {
    int status = rp_uio_release (&(handle->uio));
    if (status) {
        return (-1);
    }

    return(0);
}

int rp_clb_default(rp_clb_t *handle) {
    for (int unsigned i=0; i<handle->num_dac; i++) {
        rp_lin_default(&handle->dac[i]);
    }
    for (int unsigned i=0; i<handle->num_adc; i++) {
        rp_lin_default(&handle->adc[i]);
    }
    return(0);
}

void rp_clb_print(rp_clb_t *handle) {
    for (int unsigned i=0; i<handle->num_dac; i++) {
        rp_lin_print(&handle->dac[i]);
    }
    for (int unsigned i=0; i<handle->num_adc; i++) {
        rp_lin_print(&handle->adc[i]);
    }
}

//    def eeprom_read(self, eeprom_offset = _eeprom_offset_user):
//        # open EEPROM device
//        try:
//            eeprom_file = open(self._eeprom_device, 'rb')
//        except OSError as e:
//            raise IOError(e.errno, "Opening {}: {}".format(uio, e.strerror))
//
//        # seek to calibration data
//        try:
//            eeprom_file.seek(eeprom_offset)
//        except IOError as e:
//            raise IOError(e.errno, "Seek {}: {}".format(uio, e.strerror))
//
//        # read calibration data into a buffer
//        try:
//            buffer = eeprom_file.read(sizeof(self._eeprom_t))
//        except IOError as e:
//            raise IOError(e.errno, "Read {}: {}".format(uio, e.strerror))
//
//        # close EEPROM device
//        try:
//            eeprom_file.close()
//        except IOError as e:
//            raise IOError(e.errno, "Close {}: {}".format(uio, e.strerror))
//
//        # map buffer onto structure
//        eeprom_struct = self._eeprom_t.from_buffer_copy(buffer)
//        return eeprom_struct

static float FullScaleToVoltage(int32_t cnt) {
        if (cnt == 0)
            return(1.0);
        else
            return((float) cnt * 100.0 / ldexpf(1, 32));
}

//static int32_t FullScaleFromVoltage(float voltage) {
//        return ((int32_t) (voltage / 100.0 * ldexpf(1, 32)));
//}

int rp_clb_eeprom_parse(rp_clb_t *handle, rp_clb_eeprom_t eeprom_struct, rp_clb_float_t *clb_struct) {
    // convert EEPROM values into local float values
    for (int unsigned i=0; i<handle->num_adc; i++) {
        if (eeprom_struct.magic == CLB_MAGIC) {
            clb_struct->adc[i].lo.gain = FullScaleToVoltage(eeprom_struct.adc_lo_gain  [i]) / 20.0;
            clb_struct->adc[i].hi.gain = FullScaleToVoltage(eeprom_struct.adc_hi_gain  [i]);
        } else {
            clb_struct->adc[i].lo.gain = FullScaleToVoltage(eeprom_struct.adc_hi_gain  [i]);
            clb_struct->adc[i].hi.gain = FullScaleToVoltage(eeprom_struct.adc_lo_gain  [i]) / 20.0;
        }
        clb_struct->adc[i].lo.offset   =                    eeprom_struct.adc_lo_offset[i]  / ((1<<13)-1);
        if (eeprom_struct.magic == CLB_MAGIC) {
            clb_struct->adc[i].hi.offset =                  eeprom_struct.adc_hi_offset[i]  / ((1<<13)-1) * 20.0;
        } else {
            clb_struct->adc[i].hi.offset = clb_struct->adc[i].lo.offset;
        }
    }
    for (int unsigned i=0; i<handle->num_dac; i++) {
        clb_struct->dac[i].gain      = FullScaleToVoltage (eeprom_struct.dac_gain     [i]);
        clb_struct->dac[i].offset    =                     eeprom_struct.dac_offset   [i]  / ((1<<13)-1);
    }
    // missing magic number means a deprecated EEPROM structure was still not updated
    if (eeprom_struct.magic != CLB_MAGIC) {
        for (int unsigned i=0; i<handle->num_adc; i++) {
            clb_struct->adc[i].hi.offset = clb_struct->adc[i].lo.offset;
        }
    }

    // TODO: it probably is possible to get an error
    return (0);
}

//    def calib_show (self, clb_struct):
//        for ch in self.channels_adc:
//            print('adc[{}].lo.gain   = {}'.format(ch, clb_struct.adc[ch].lo.gain))
//            print('adc[{}].hi.gain   = {}'.format(ch, clb_struct.adc[ch].hi.gain))
//            print('adc[{}].lo.offset = {}'.format(ch, clb_struct.adc[ch].lo.offset))
//            print('adc[{}].hi.offset = {}'.format(ch, clb_struct.adc[ch].hi.offset))
//        for ch in self.channels_dac:
//            print('dac[{}].gain      = {}'.format(ch, clb_struct.dac[ch].gain))
//            print('dac[{}].offset    = {}'.format(ch, clb_struct.dac[ch].offset))
//
//int rp_clb_apply (self, clb_struct, adc_range = ['lo', 'lo']):
//    for ch in self.channels_adc:
//        if   (adc_range[ch] == 'lo'):
//            self.adc[ch].gain   = clb_struct.adc[ch].lo.gain
//            self.adc[ch].offset = clb_struct.adc[ch].lo.offset
//        elif (adc_range[ch] == 'hi'):
//            self.adc[ch].gain   = clb_struct.adc[ch].hi.gain
//            self.adc[ch].offset = clb_struct.adc[ch].hi.offset
//        else:
//            raise ValueError("ADC range can be one of ['lo', 'hi'].")
//    for ch in self.channels_dac:
//        self.dac[ch].gain   = clb_struct.dac[ch].gain
//        self.dac[ch].offset = clb_struct.dac[ch].offset
